lib/services/watcher_test.go

```diff
@@ -0,0 +1,203 @@
+		t.Fatal("Watcher has unexpectedly exited.")
+	case <-time.After(2 * time.Second):
+		t.Fatal("Timeout waiting for the update event.")
+	}
+
+	// Delete the first proxy.
+	require.NoError(t, presence.DeleteProxy(proxy.GetName()))
+
+	// Watcher should detect the proxy list change.
+	select {
+	case changeset := <-w.ProxiesC:
+		require.Len(t, changeset, 1)
+		require.Empty(t, resourceDiff(changeset[0], proxy2))
+	case <-w.Done():
+		t.Fatal("Watcher has unexpectedly exited.")
+	case <-time.After(2 * time.Second):
+		t.Fatal("Timeout waiting for the update event.")
+	}
+}
+
+func newProxyServer(t *testing.T, name, addr string) types.Server {
+	s, err := types.NewServer(name, types.KindProxy, types.ServerSpecV2{
+		Addr:       addr,
+		PublicAddr: addr,
+	})
+	require.NoError(t, err)
+	return s
+}
+
+func TestLockWatcher(t *testing.T) {
+	t.Parallel()
+	ctx := context.Background()
+	clock := clockwork.NewFakeClock()
+
+	bk, err := lite.NewWithConfig(ctx, lite.Config{
+		Path:             t.TempDir(),
+		PollStreamPeriod: 200 * time.Millisecond,
+		Clock:            clock,
+	})
+	require.NoError(t, err)
+
+	type client struct {
+		services.Access
+		types.Events
+	}
+
+	access := local.NewAccessService(bk)
+	w, err := services.NewLockWatcher(ctx, services.LockWatcherConfig{
+		ResourceWatcherConfig: services.ResourceWatcherConfig{
+			Component:   "test",
+			RetryPeriod: 200 * time.Millisecond,
+			Client: &client{
+				Access: access,
+				Events: local.NewEventsService(bk, nil),
+			},
+			Clock: clock,
+		},
+	})
+	require.NoError(t, err)
+	t.Cleanup(w.Close)
+
+	// Subscribe to lock watcher updates.
+	target := types.LockTarget{Node: "node"}
+	require.Nil(t, w.GetSomeLockInForce(target))
+	sub, err := w.Subscribe(ctx, target)
+	require.NoError(t, err)
+	t.Cleanup(func() { require.NoError(t, sub.Close()) })
+
+	// Add an *expired* lock matching the subscription target.
+	pastTime := clock.Now().Add(-time.Minute)
+	lock, err := types.NewLock("lock", types.LockSpecV2{
+		Target:  target,
+		Expires: &pastTime,
+	})
+	require.NoError(t, err)
+	require.NoError(t, access.UpsertLock(ctx, lock))
+	select {
+	case event := <-sub.Events():
+		t.Fatalf("Unexpected event: %v.", event)
+	case <-sub.Done():
+		t.Fatal("Lock watcher subscription has unexpectedly exited.")
+	case <-time.After(2 * time.Second):
+	}
+	require.Nil(t, w.GetSomeLockInForce(target))
+
+	// Update the latter lock so it becomes in force.
+	futureTime := clock.Now().Add(time.Minute)
+	lock.SetLockExpiry(&futureTime)
+	require.NoError(t, access.UpsertLock(ctx, lock))
+	select {
+	case event := <-sub.Events():
+		receivedLock, ok := event.Resource.(types.Lock)
+		require.True(t, ok)
+		require.Empty(t, resourceDiff(receivedLock, lock))
+	case <-sub.Done():
+		t.Fatal("Lock watcher subscription has unexpectedly exited.")
+	case <-time.After(2 * time.Second):
+		t.Fatal("Timeout waiting for the update event.")
+	}
+	require.Empty(t, resourceDiff(w.GetSomeLockInForce(target), lock))
+}
```

Comment:
Can you add cases for adding a second lock and deleting a lock like the proxy watcher tests?

Reply:
`LockWatcher` isn't exactly analogous to `ProxyWatcher` since lock deletion is _not_ broadcast to the fanout subscribers. That's why the `LockWatcher` test doesn't feature second-addition/deletion.

Reply:
I'm still not quite decided about this though. Broadcasting lock deletions might be preferable to do (even if not strictly needed for my current use cases) in which case I should update the test as you suggested.